File: frontend/apps/web-antd/src/api/admin/periodic-task.ts
/**
 * Periodic task API / 定时任务 API
 * Backend: /admin/periodic-tasks/*
 */
import type {
  PeriodicTaskFormData,
  PeriodicTaskInfo,
  PeriodicTaskListParams,
  PeriodicTaskListResponse,
} from '#/api/shared/periodic-task-types';
import type { ApiRequestOptions } from '#/utils/request';

import { $t, $te } from '#/locales';
import { requestClient } from '#/utils/request';

// ============================================================
// Type definitions (imported from shared) / 类型定义（从 shared 导入）
// ============================================================

export type {
  PeriodicTaskFormData,
  PeriodicTaskInfo,
  PeriodicTaskListParams,
  PeriodicTaskListResponse,
} from '#/api/shared/periodic-task-types';

/** Periodic task info (backend raw snake_case) / 定时任务信息（后端原始） */
export interface PeriodicTaskInfoRaw {
  id: number;
  name: string;
  task_path: string;
  definition_type?: string;
  schedule_type: string;
  cron_expression: null | string;
  interval_seconds: null | number;
  is_active: boolean;
  last_run_at: null | string;
  next_run_at: null | string;
  description: null | string;
  plugin_enabled?: boolean;
  plugin_name?: null | string;
  created_at: string;
  scope: null | string;
  owner_tenant_id: null | number;
  assigned_tenant_ids?: number[];
  assigned_tenant_names?: string[];
  binding_count?: number;
  binding_required?: boolean;
  binding_configured?: boolean;
  tenant_access_mode?: string;
  binding_summary?: null | string;
  is_locked: boolean;
  is_editable: boolean;
  default_priority?: null | number;
  required_feature_codes?: string[];
  required_plugin_names?: string[];
  max_retries: number;
  retry_delay: number;
  timeout: null | number;
  notify_on_failure: boolean;
  notify_emails: null | string;
}

export interface PeriodicTaskBindingInfo {
  id: number;
  tenantId: number;
  tenantName: null | string;
  isEnabled: boolean;
  disabledReason: null | string;
  scheduleTypeOverride: null | string;
  cronExpressionOverride: null | string;
  intervalSecondsOverride: null | number;
  kwargsOverride: null | Record<string, unknown>;
  configOverride: null | Record<string, unknown>;
  effectiveScheduleType: null | string;
  effectiveCronExpression: null | string;
  effectiveIntervalSeconds: null | number;
  lastRunAt: null | string;
  nextRunAt: null | string;
}

export interface PeriodicTaskBindingInfoRaw {
  id: number;
  tenant_id: number;
  tenant_name?: null | string;
  is_enabled?: boolean;
  disabled_reason?: null | string;
  schedule_type_override?: null | string;
  cron_expression_override?: null | string;
  interval_seconds_override?: null | number;
  kwargs_override?: null | Record<string, unknown>;
  config_override?: null | Record<string, unknown>;
  effective_schedule_type?: null | string;
  effective_cron_expression?: null | string;
  effective_interval_seconds?: null | number;
  last_run_at?: null | string;
  next_run_at?: null | string;
}

export interface PeriodicTaskBindingUpdatePayload {
  isEnabled?: boolean;
  disabledReason?: null | string;
  scheduleTypeOverride?: null | string;
  cronExpressionOverride?: null | string;
  intervalSecondsOverride?: null | number;
  kwargsOverride?: null | Record<string, unknown>;
  configOverride?: null | Record<string, unknown>;
}

export interface PeriodicTaskBindingSyncItemPayload extends PeriodicTaskBindingUpdatePayload {
  tenantId: number;
}

export interface PeriodicTaskBindingSyncPayload {
  bindings?: PeriodicTaskBindingSyncItemPayload[];
  scope?: null | string;
  tenant_ids?: number[];
}

function getTaskLeafCandidates(raw: PeriodicTaskInfoRaw): string[] {
  const candidates = [raw.task_path, raw.name]
    .filter(Boolean)
    .flatMap((value) => {
      const trimmed = value.trim();
      const leaf = trimmed.split('.').at(-1) || trimmed;
      return [trimmed, leaf];
    });

  return [...new Set(candidates)];
}

function translateSystemTaskName(raw: PeriodicTaskInfoRaw): string {
  if (raw.definition_type === 'plugin') {
    return raw.name;
  }

  for (const candidate of getTaskLeafCandidates(raw)) {
    const key = `admin.system.taskLog.taskNames.${candidate}`;
    if ($te(key)) {
      return $t(key);
    }
  }

  return raw.name;
}

function translateSystemTaskDescription(
  raw: PeriodicTaskInfoRaw,
): null | string {
  if (raw.definition_type === 'plugin' || !raw.description) {
    return raw.description;
  }

  for (const candidate of getTaskLeafCandidates(raw)) {
    const key = `admin.system.periodicTask.taskDescriptions.${candidate}`;
    if ($te(key)) {
      return $t(key);
    }
  }

  return raw.description;
}

// ============================================================
// Transform functions / 转换函数
// ============================================================

function transformPeriodicTaskInfo(raw: PeriodicTaskInfoRaw): PeriodicTaskInfo {
  return {
    id: raw.id,
    name: translateSystemTaskName(raw),
    taskPath: raw.task_path,
    definitionType: raw.definition_type ?? 'system',
    scheduleType: raw.schedule_type,
    cronExpression: raw.cron_expression,
    intervalSeconds: raw.interval_seconds,
    isActive: raw.is_active,
    lastRunAt: raw.last_run_at,
    nextRunAt: raw.next_run_at,
    description: translateSystemTaskDescription(raw),
    pluginEnabled: raw.plugin_enabled ?? true,
    pluginName: raw.plugin_name ?? null,
    createdAt: raw.created_at,
    scope: raw.scope,
    tenantId: raw.owner_tenant_id,
    assignedTenantIds: raw.assigned_tenant_ids ?? [],
    assignedTenantNames: raw.assigned_tenant_names ?? [],
    bindingCount: raw.binding_count ?? 0,
    bindingRequired: raw.binding_required ?? false,
    bindingConfigured: raw.binding_configured ?? true,
    tenantAccessMode: raw.tenant_access_mode ?? 'none',
    bindingSummary: raw.binding_summary ?? null,
    isLocked: raw.is_locked,
    isEditable: raw.is_editable,
    defaultPriority: raw.default_priority ?? null,
    requiredFeatureCodes: raw.required_feature_codes ?? [],
    requiredPluginNames: raw.required_plugin_names ?? [],
    maxRetries: raw.max_retries,
    retryDelay: raw.retry_delay,
    timeout: raw.timeout,
    notifyOnFailure: raw.notify_on_failure,
    notifyEmails: raw.notify_emails,
  };
}

function transformPeriodicTaskBindingInfo(
  raw: PeriodicTaskBindingInfoRaw,
): PeriodicTaskBindingInfo {
  return {
    id: raw.id,
    tenantId: raw.tenant_id,
    tenantName: raw.tenant_name ?? null,
    isEnabled: raw.is_enabled ?? true,
    disabledReason: raw.disabled_reason ?? null,
    scheduleTypeOverride: raw.schedule_type_override ?? null,
    cronExpressionOverride: raw.cron_expression_override ?? null,
    intervalSecondsOverride: raw.interval_seconds_override ?? null,
    kwargsOverride: raw.kwargs_override ?? null,
    configOverride: raw.config_override ?? null,
    effectiveScheduleType: raw.effective_schedule_type ?? null,
    effectiveCronExpression: raw.effective_cron_expression ?? null,
    effectiveIntervalSeconds: raw.effective_interval_seconds ?? null,
    lastRunAt: raw.last_run_at ?? null,
    nextRunAt: raw.next_run_at ?? null,
  };
}

function compactPayload<T extends Record<string, unknown>>(payload: T): T {
  return Object.fromEntries(
    Object.entries(payload).filter(([, value]) => value !== undefined),
  ) as T;
}

function toBindingRawPayload(payload: PeriodicTaskBindingUpdatePayload) {
  return compactPayload({
    is_enabled: payload.isEnabled,
    disabled_reason: payload.disabledReason,
    schedule_type_override: payload.scheduleTypeOverride,
    cron_expression_override: payload.cronExpressionOverride,
    interval_seconds_override: payload.intervalSecondsOverride,
    kwargs_override: payload.kwargsOverride,
    config_override: payload.configOverride,
  });
}

function toBindingSyncRawPayload(payload: PeriodicTaskBindingSyncItemPayload) {
  return {
    tenant_id: payload.tenantId,
    ...toBindingRawPayload(payload),
  };
}

// ============================================================
// API functions / API 接口
// ============================================================

const API_PREFIX = '/admin/periodic-tasks';

/**
 * Get periodic task list / 获取定时任务列表
 * GET /admin/periodic-tasks
 */
export async function getPeriodicTaskListApi(
  params?: PeriodicTaskListParams,
  options?: ApiRequestOptions,
): Promise<PeriodicTaskListResponse> {
  const response = await requestClient.get<{
    items: PeriodicTaskInfoRaw[];
    page: number;
    page_size: number;
    total: number;
  }>(API_PREFIX, { params, ...options });

  return {
    items: response.items.map((item) => transformPeriodicTaskInfo(item)),
    total: response.total,
    page: response.page,
    page_size: response.page_size,
  };
}

/**
 * Get periodic task detail / 获取定时任务详情
 * GET /admin/periodic-tasks/{id}
 */
export async function getPeriodicTaskDetailApi(
  id: number,
  options?: ApiRequestOptions,
): Promise<PeriodicTaskInfo> {
  const raw = await requestClient.get<PeriodicTaskInfoRaw>(
    `${API_PREFIX}/${id}`,
    options,
  );
  return transformPeriodicTaskInfo(raw);
}

/**
 * Create periodic task / 创建定时任务
 * POST /admin/periodic-tasks
 */
export async function createPeriodicTaskApi(
  data: PeriodicTaskFormData,
  options?: ApiRequestOptions,
): Promise<PeriodicTaskInfo> {
  const raw = await requestClient.post<PeriodicTaskInfoRaw>(
    API_PREFIX,
    data,
    options,
  );
  return transformPeriodicTaskInfo(raw);
}

/**
 * Update periodic task / 更新定时任务
 * PUT /admin/periodic-tasks/{id}
 */
export async function updatePeriodicTaskApi(
  id: number,
  data: Partial<PeriodicTaskFormData>,
  options?: ApiRequestOptions,
): Promise<PeriodicTaskInfo> {
  const raw = await requestClient.put<PeriodicTaskInfoRaw>(
    `${API_PREFIX}/${id}`,
    data,
    options,
  );
  return transformPeriodicTaskInfo(raw);
}

/**
 * Delete periodic task / 删除定时任务
 * DELETE /admin/periodic-tasks/{id}
 */
export async function deletePeriodicTaskApi(
  id: number,
  options?: ApiRequestOptions,
): Promise<void> {
  await requestClient.delete(`${API_PREFIX}/${id}`, options);
}

/**
 * Enable/disable periodic task / 启用/禁用定时任务
 * POST /admin/periodic-tasks/{id}/toggle
 */
export async function togglePeriodicTaskApi(
  id: number,
  isActive: boolean,
  options?: ApiRequestOptions,
): Promise<PeriodicTaskInfo> {
  const raw = await requestClient.post<PeriodicTaskInfoRaw>(
    `${API_PREFIX}/${id}/toggle`,
    { is_active: isActive },
    options,
  );
  return transformPeriodicTaskInfo(raw);
}

/**
 * Manually trigger periodic task / 手动触发定时任务
 * POST /admin/periodic-tasks/{id}/trigger
 */
export async function triggerPeriodicTaskApi(
  id: number,
  options?: ApiRequestOptions,
): Promise<{ triggeredTaskId: string }> {
  const raw = await requestClient.post<{ triggered_task_id: string }>(
    `${API_PREFIX}/${id}/trigger`,
    {},
    options,
  );
  return { triggeredTaskId: raw.triggered_task_id };
}

/**
 * Get periodic task tenant bindings / 获取定时任务企业绑定
 * GET /admin/periodic-tasks/{id}/bindings
 */
export async function getPeriodicTaskBindingsApi(
  id: number,
  options?: ApiRequestOptions,
): Promise<PeriodicTaskBindingInfo[]> {
  const response = await requestClient.get<PeriodicTaskBindingInfoRaw[]>(
    `${API_PREFIX}/${id}/bindings`,
    options,
  );
  return response.map((item) => transformPeriodicTaskBindingInfo(item));
}

/**
 * Sync periodic task tenant bindings / 同步定时任务企业绑定
 * PUT /admin/periodic-tasks/{id}/bindings
 */
export async function syncPeriodicTaskBindingsApi(
  id: number,
  payload: PeriodicTaskBindingSyncPayload,
  options?: ApiRequestOptions,
): Promise<{ added: number; reenabled: number; removed: number }> {
  const body = compactPayload({
    scope: payload.scope,
    tenant_ids: payload.tenant_ids ?? [],
    bindings: payload.bindings?.map((item) => toBindingSyncRawPayload(item)),
  });
  return await requestClient.put<{
    added: number;
    reenabled: number;
    removed: number;
  }>(`${API_PREFIX}/${id}/bindings`, body, options);
}

/**
 * Update one tenant binding / 更新单个企业绑定
 * PATCH /admin/periodic-tasks/{id}/bindings/{tenant_id}
 */
export async function updatePeriodicTaskBindingApi(
  id: number,
  tenantId: number,
  payload: PeriodicTaskBindingUpdatePayload,
  options?: ApiRequestOptions,
): Promise<PeriodicTaskBindingInfo> {
  const raw = await requestClient.patch<PeriodicTaskBindingInfoRaw>(
    `${API_PREFIX}/${id}/bindings/${tenantId}`,
    toBindingRawPayload(payload),
    options,
  );
  return transformPeriodicTaskBindingInfo(raw);
}

File: backend/app/schemas/system/periodic_task.py
"""
定时任务相关 Schema / Periodic Task Schema

定义定时任务管理 API 的请求和响应数据结构
Defines periodic task management API request and response data structures.
"""

from datetime import datetime

from pydantic import Field, field_validator, model_validator

from app.core.base_schema import BaseSchema
from app.enums.common import ResourceScopeEnum

ALLOWED_PERIODIC_TASK_SCOPES = {
    ResourceScopeEnum.GLOBAL_SHARED.value,
    ResourceScopeEnum.ADMIN_ONLY.value,
    ResourceScopeEnum.ALL_TENANTS.value,
    ResourceScopeEnum.ADMIN_AND_SELECTED_TENANTS.value,
    ResourceScopeEnum.SELECTED_TENANTS.value,
}


class PeriodicTaskResponse(BaseSchema):
    """定时任务响应 / Periodic task response."""

    id: int = Field(..., description="ID")
    name: str = Field(..., description="任务名称")
    definition_type: str = Field("system", description="任务定义类型")
    task_path: str = Field(..., description="任务路径")
    schedule_type: str = Field(..., description="调度类型")
    cron_expression: str | None = Field(None, description="Cron 表达式")
    interval_seconds: int | None = Field(None, description="间隔秒数")
    is_active: bool = Field(True, description="是否启用")
    last_run_at: datetime | None = Field(None, description="上次执行时间")
    next_run_at: datetime | None = Field(None, description="下次执行时间")
    description: str | None = Field(None, description="任务描述")
    plugin_name: str | None = Field(None, description="插件名称（插件任务时返回）")
    plugin_enabled: bool = Field(True, description="关联插件当前是否已启用")
    scope: str = Field("admin_only", description="资源作用域 ResourceScopeEnum")
    owner_tenant_id: int | None = Field(None, description="归属企业ID（平台级为 NULL）")
    assigned_tenant_ids: list[int] = Field(
        default_factory=list,
        description="当前已绑定的企业 ID 列表",
    )
    assigned_tenant_names: list[str] = Field(
        default_factory=list,
        description="当前已绑定的企业名称列表",
    )
    binding_count: int = Field(0, description="已绑定企业数量")
    binding_required: bool = Field(False, description="当前 scope 是否要求选择企业")
    binding_configured: bool = Field(True, description="当前分发模式是否已完成绑定")
    tenant_access_mode: str = Field(
        "none", description="企业覆盖模式 none/all/selected"
    )
    binding_summary: str | None = Field(None, description="绑定摘要")
    is_locked: bool = Field(False, description="是否禁止删除")
    is_editable: bool = Field(True, description="是否允许编辑")
    default_priority: int | None = Field(
        None,
        ge=0,
        le=9,
        description="默认 Celery broker 优先级（0-9）",
    )
    required_feature_codes: list[str] = Field(
        default_factory=list,
        description="要求的企业套餐特性代码",
    )
    required_plugin_names: list[str] = Field(
        default_factory=list,
        description="要求的企业可用插件名称",
    )
    max_retries: int = Field(0, description="最大重试次数")
    retry_delay: int = Field(60, description="重试间隔（秒）")
    timeout: int = Field(3600, description="执行超时（秒）")
    notify_on_failure: bool = Field(False, description="失败时是否通知")
    notify_emails: str | None = Field(None, description="通知邮箱列表")
    created_at: datetime = Field(..., description="创建时间")


class PeriodicTaskCreateRequest(BaseSchema):
    """创建定时任务请求 / Create periodic task request."""

    name: str = Field(..., min_length=1, max_length=255, description="任务名称")
    task_path: str = Field(..., min_length=1, description="任务路径")
    schedule_type: str = Field("interval", description="调度类型（cron/interval）")
    cron_expression: str | None = Field(None, description="Cron 表达式")
    interval_seconds: int | None = Field(
        None, ge=10, description="间隔秒数（最小10秒）"
    )
    args: dict | None = Field(None, description="位置参数")
    kwargs: dict | None = Field(None, description="关键字参数")
    is_active: bool = Field(True, description="是否启用")
    description: str | None = Field(None, description="任务描述")
    scope: str = Field("admin_only", description="资源作用域 ResourceScopeEnum（五类）")
    owner_tenant_id: int | None = Field(
        None, description="归属企业ID（单企业任务时填写）"
    )
    tenant_ids: list[int] = Field(default_factory=list, description="分发企业 ID 列表")
    default_priority: int | None = Field(
        None,
        ge=0,
        le=9,
        description="默认 Celery broker 优先级（0-9）",
    )
    required_feature_codes: list[str] = Field(
        default_factory=list,
        description="要求的企业套餐特性代码",
    )
    required_plugin_names: list[str] = Field(
        default_factory=list,
        description="要求的企业可用插件名称",
    )
    max_retries: int = Field(0, ge=0, le=10, description="最大重试次数")
    retry_delay: int = Field(60, ge=1, le=3600, description="重试间隔（秒）")
    timeout: int = Field(3600, ge=10, le=86400, description="执行超时（秒）")
    notify_on_failure: bool = Field(False, description="失败时是否通知")
    notify_emails: str | None = Field(None, description="通知邮箱列表（逗号分隔）")

    @field_validator("scope")
    @classmethod
    def validate_scope_value(cls, value: str) -> str:
        if value not in ALLOWED_PERIODIC_TASK_SCOPES:
            raise ValueError("invalid scope")
        return value

    @model_validator(mode="after")
    def validate_scope_owner(self):
        if self.scope in (
            ResourceScopeEnum.ADMIN_ONLY.value,
            ResourceScopeEnum.GLOBAL_SHARED.value,
            ResourceScopeEnum.ALL_TENANTS.value,
        ):
            self.owner_tenant_id = None
        if self.scope not in (
            ResourceScopeEnum.SELECTED_TENANTS.value,
            ResourceScopeEnum.ADMIN_AND_SELECTED_TENANTS.value,
        ):
            self.tenant_ids = []
        return self


class PeriodicTaskUpdateRequest(BaseSchema):
    """更新定时任务请求 / Update periodic task request."""

    name: str | None = Field(None, min_length=1, max_length=255, description="任务名称")
    task_path: str | None = Field(None, min_length=1, description="任务路径")
    schedule_type: str | None = Field(None, description="调度类型")
    cron_expression: str | None = Field(None, description="Cron 表达式")
    interval_seconds: int | None = Field(None, ge=10, description="间隔秒数")
    args: dict | None = Field(None, description="位置参数")
    kwargs: dict | None = Field(None, description="关键字参数")
    is_active: bool | None = Field(None, description="是否启用")
    description: str | None = Field(None, description="任务描述")
    scope: str | None = Field(None, description="资源作用域")
    owner_tenant_id: int | None = Field(None, description="归属企业ID")
    tenant_ids: list[int] | None = Field(None, description="分发企业 ID 列表")
    default_priority: int | None = Field(
        None,
        ge=0,
        le=9,
        description="默认 Celery broker 优先级（0-9）",
    )
    required_feature_codes: list[str] | None = Field(
        None,
        description="要求的企业套餐特性代码",
    )
    required_plugin_names: list[str] | None = Field(
        None,
        description="要求的企业可用插件名称",
    )
    max_retries: int | None = Field(None, ge=0, le=10, description="最大重试次数")
    retry_delay: int | None = Field(None, ge=1, le=3600, description="重试间隔（秒）")
    timeout: int | None = Field(None, ge=10, le=86400, description="执行超时（秒）")
    notify_on_failure: bool | None = Field(None, description="失败时是否通知")
    notify_emails: str | None = Field(None, description="通知邮箱列表")

    @field_validator("scope")
    @classmethod
    def validate_scope_value(cls, value: str | None) -> str | None:
        if value is not None and value not in ALLOWED_PERIODIC_TASK_SCOPES:
            raise ValueError("invalid scope")
        return value

    @model_validator(mode="after")
    def validate_selected_scope_tenants(self):
        if self.scope in (
            ResourceScopeEnum.ADMIN_ONLY.value,
            ResourceScopeEnum.GLOBAL_SHARED.value,
            ResourceScopeEnum.ALL_TENANTS.value,
        ):
            self.owner_tenant_id = None
        return self


class PeriodicTaskToggleRequest(BaseSchema):
    """启用/禁用请求 / Enable/disable request."""

    is_active: bool = Field(..., description="是否启用")


class PeriodicTaskBindingResponse(BaseSchema):
    """定时任务企业绑定响应 / Periodic task tenant binding response."""

    id: int = Field(..., description="绑定 ID")
    tenant_id: int = Field(..., description="企业 ID")
    tenant_name: str | None = Field(None, description="企业名称")
    is_enabled: bool = Field(True, description="是否启用")
    disabled_reason: str | None = Field(None, description="禁用原因")
    schedule_type_override: str | None = Field(None, description="覆盖调度类型")
    cron_expression_override: str | None = Field(None, description="覆盖 Cron 表达式")
    interval_seconds_override: int | None = Field(None, description="覆盖间隔秒数")
    config_override: dict | None = Field(None, description="覆盖配置")
    args_override: dict | list | None = Field(None, description="覆盖位置参数")
    kwargs_override: dict | None = Field(None, description="覆盖关键字参数")
    effective_schedule_type: str | None = Field(None, description="生效调度类型")
    effective_cron_expression: str | None = Field(None, description="生效 Cron 表达式")
    effective_interval_seconds: int | None = Field(None, description="生效间隔秒数")
    last_run_at: datetime | None = Field(None, description="上次执行时间")
    next_run_at: datetime | None = Field(None, description="下次执行时间")


class PeriodicTaskBindingUpdateRequest(BaseSchema):
    """单条企业绑定更新请求 / Single tenant binding update request."""

    is_enabled: bool | None = Field(None, description="是否启用")
    disabled_reason: str | None = Field(None, max_length=500, description="禁用原因")
    schedule_type_override: str | None = Field(None, description="覆盖调度类型")
    cron_expression_override: str | None = Field(None, description="覆盖 Cron 表达式")
    interval_seconds_override: int | None = Field(
        None,
        ge=10,
        description="覆盖间隔秒数",
    )
    config_override: dict | None = Field(None, description="覆盖配置")
    args_override: dict | list | None = Field(None, description="覆盖位置参数")
    kwargs_override: dict | None = Field(None, description="覆盖关键字参数")

    @field_validator("schedule_type_override")
    @classmethod
    def validate_schedule_type_override(cls, value: str | None) -> str | None:
        if value is not None and value not in {"cron", "interval"}:
            raise ValueError("invalid schedule_type_override")
        return value


class PeriodicTaskBindingSyncItem(PeriodicTaskBindingUpdateRequest):
    """批量同步中的单条企业绑定配置 / Tenant binding config in batch sync."""

    tenant_id: int = Field(..., description="企业 ID")


class PeriodicTaskBindingSyncRequest(BaseSchema):
    """定时任务企业绑定同步请求 / Periodic task tenant binding sync request."""

    tenant_ids: list[int] = Field(default_factory=list, description="企业 ID 列表")
    bindings: list[PeriodicTaskBindingSyncItem] = Field(
        default_factory=list,
        description="企业绑定覆盖配置列表",
    )
    scope: str | None = Field(
        None,
        description="同步后的目标作用域；不传时按是否有企业绑定自动推导",
    )

    @field_validator("scope")
    @classmethod
    def validate_scope_value(cls, value: str | None) -> str | None:
        if value is not None and value not in ALLOWED_PERIODIC_TASK_SCOPES:
            raise ValueError("invalid scope")
        return value

    @model_validator(mode="after")
    def normalize_tenant_ids_for_non_explicit_scope(self):
        if self.scope is not None and self.scope not in (
            ResourceScopeEnum.SELECTED_TENANTS.value,
            ResourceScopeEnum.ADMIN_AND_SELECTED_TENANTS.value,
            ResourceScopeEnum.ALL_TENANTS.value,
        ):
            self.tenant_ids = []
        return self


__all__ = [
    "PeriodicTaskResponse",
    "PeriodicTaskCreateRequest",
    "PeriodicTaskUpdateRequest",
    "PeriodicTaskToggleRequest",
    "PeriodicTaskBindingResponse",
    "PeriodicTaskBindingUpdateRequest",
    "PeriodicTaskBindingSyncItem",
    "PeriodicTaskBindingSyncRequest",
]
